Add download queue data button

jackieo5023 opened the pull request:
This PR:
* Adds a button and endpoint to download course queue data as a csv on the course page
    * This is only viewable to course staff, which I have tested
* Changed a comment that was probably not changed after copy/paste to say "remove" instead of "add"

![image](https://user-images.githubusercontent.com/32113753/66261553-2d707980-e795-11e9-8f21-b02f9c5d7a10.png)

[Here](https://github.com/illinois/queue/files/3693750/queueData.txt) is an example of what the download would look like given fake data.  Unfortunately, I couldn't upload to GitHub as a csv.
Here is the SQL query that this was based off of:
```
SELECT
  questions.id AS id,
  courses.name AS CourseName,
  queues.name AS QueueName,
  u_asked.netid AS AskedBy_netid, u_asked.universityName AS AskedBy_RealName,
  u_answered.netid AS AnsweredBy_netid, u_answered.universityName AS AnsweredBy_RealName,
  topic,
  CONVERT_TZ(enqueueTime, '+0:00', 'US/Central') AS enqueueTime,
  CONVERT_TZ(dequeueTime, '+0:00', 'US/Central') AS dequeueTime,
  CONVERT_TZ(answerStartTime, '+0:00', 'US/Central') AS answerStartTime,
  CONVERT_TZ(answerFinishTime, '+0:00', 'US/Central') AS answerFinishTime,
  comments, preparedness,
  questions.location AS UserLocation,
  queues.location AS QueueLocation,
  CONVERT_TZ(queues.createdAt, '+0:00', 'US/Central') AS Queue_CreatedAt,
  queueId, courseId
FROM questions
  INNER JOIN queues ON questions.queueId = queues.id
  INNER JOIN courses ON queues.courseId = courses.id
  INNER JOIN users u_asked ON questions.askedById = u_asked.id
  LEFT JOIN users u_answered ON questions.answeredById = u_answered.id
WHERE courseId=6
ORDER BY enqueueTime DESC
```

nwalters512 commented on src/api/courses.js:
```diff
@@ -69,6 +69,124 @@ router.get(
 
+// Get course queue data
+router.get(
+  '/:courseId/data',
```
I would consider scoping this to something like `data/questions` or the like - we might want more "data" endpoints in the future. One that immediately comes to mind is a `data/staff` endpoint that could return all join/leave records of queue staff.

nwalters512 commented on src/pages/course.js:
```diff
@@ -32,6 +33,76 @@ const Course = props => {
     })
   }, [props.courseId])
 
+  const getFlattenedData = resData => {
```
It's not clear to me why this is being done on the client, or even what exactly this is doing. Could you elaborate on the former, and (if this stays on the client) can you add some comments explaining what this does?

Also, it looks like this function doesn't close over any values in the component - this could be moved outside the component to keep the component logic itself clear.

jackieo5023 replied:
I can definitely move this to the api, that would make more sense.  As for what it does - from what I found, Sequelize doesn't have a way to not "nest" objects... So if I'm querying for a question and want to include info about the queue (like name, id) the question belongs to, the queue fields (name, id) will be nested under a larger "queue" object.  So this just flattens all of the questions returned.

nwalters512 replied:
Makes sense - thanks for clarifying! This feels like one of those cases where Sequelize gets in our way much more than it helps us - raw SQL seems like it would be so much better here 🙂

james9909 commented:
Adding onto Nathan's comments, we should also add some tests for this new endpoint

nwalters512 commented on src/api/courses.js:
```diff
@@ -69,6 +69,124 @@ router.get(
+              ['universityName', 'AskedBy_RealName'],
+            ],
+            exclude: [
+              'createdAt',
```
Is excluding all of these actually required? This seems somewhat brittle if we were to add more fields to users in the future; ditto for other models here.

jackieo5023 replied:
There seemed to be some weirdness with the 'name' field.  If I included just netid and universityName, additional fields called `universityName` and `preferredName` would show up in the response.

jackieo5023 replied:
I can add a test to make sure if fields are added, that it fails if they are not excluded.

nwalters512 commented on src/api/courses.js:
```diff
@@ -69,6 +69,124 @@ router.get(
+          attributes: {
+            include: [
+              ['netid', 'AskedBy_netid'],
+              ['universityName', 'AskedBy_RealName'],
```
Calling something a "real" name has certain unwanted connotations. For instance, I have trans friends whose university name is not their real name; there are also students who use a different name day-to-day from what's in Enterprise. I think there are a couple potential changes here:

* calling this `UniversityName`
* the above, but also exporting the `PreferredName` field
* doing what the UI/API already do: taking the preferred name if it's specified and falling back to the name provided by Shib if they aren't
* omitting the name entirely

I think my personal preference would be the third option: it's most consistent with how names are treated everywhere else in the queue. There is the potential for abuse (I can set my preferred name to an arbitrary string), but the nedid can always be cross-referenced with a roster or the university directory if there are doubts about a user's identity.

jackieo5023 replied:
I'll play around with the options you listed here - maybe one of them can solve the weird issue mentioned above.

jackieo5023 commented:
@nwalters512 as for your point about downloads being in another section, I think at this point @wadefagen just wants downloads to be pushed out as quick as possible - so I guess think of this as a temporary solution, and later I can make it nicer to be able to download per queue as well.

jackieo5023 commented:
Now ready for review again.

* Sequelize is weird in that if you specifically include an attribute in a related model, it doesn't exclude other attributes.  So in the `getColumns` function, I whitelist the fields we want to display.
* Added tests for the endpoint.
* RealName is changed to UniversityName
* Moved all logic to api except for the download

[Here](https://github.com/illinois/queue/files/3711118/queueData.txt) is an updated version of an example csv.

jackieo5023 commented:
Tested on staging.  Everything appears to be working properly and correctly (times are verified to be in CST as well).

nwalters512 commented:
Going to give this a final look over!

nwalters512 commented on src/api/courses.js:
```diff
@@ -4,13 +4,87 @@ const router = require('express').Router({
 const safeAsync = require('../middleware/safeAsync')
 
+// Helper for getting course question data
+const getColumns = questions => {
```
It's a bit unclear to me what this function does - could you add a comment describing specifically what it returns?

nwalters512 commented on src/api/courses.js:
```diff
@@ -69,6 +143,74 @@ router.get(
+      raw: true,
+    })
+
+    res.send(getCsv(questions))
```
The response should have a `Content-Type` of `text/csv`. It could also include a `Content-Disposition` specifying that the response is an attachment and should be downloaded instead of being displayed in the browser. More details: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition

nwalters512 commented on src/pages/course.js:
```diff
@@ -32,6 +33,23 @@ const Course = props => {
     })
   }, [props.courseId])
 
+  const handleFetchQueueData = () => {
```
If you use the `Content-Disposition` method mentioned in my other comment, you should be able to avoid doing all this stuff on the client and let the browser handle downloading for you.

nwalters512 commented on .eslintrc.js:
```diff
@@ -1,6 +1,9 @@
 module.exports = {
   // parser: 'babel-eslint',
   parser: '@typescript-eslint/parser',
+  env: {
```
Nit: I'd prefer an explicit `/* eslint-env browser */` comment on files where DOM APIs are used; that will allow ESLint to warn us if, for instance, someone accidentally tries to use a browser API in server code.

nwalters512 commented on src/api/courses.js:
```diff
@@ -4,13 +4,87 @@ const router = require('express').Router({
+          const formattedTime =
+            time !== null
+              ? moment
+                  .tz(time, 'YYYY-MM-DD HH:mm:ss.SSS Z', 'US/Central')
```
Thought: this may be undesirable if/when we roll out to schools like UBC that aren't in the central timezone. When that happens, we should either a) report times in UTC and allow consumers of this data to transform it as needed, or b) associate some concept of timezone with either courses or institutions (one institutions are an entity that we have). I think I'd prefer the latter; at any rate, that's outside the scope of this PR.

nwalters512 commented on src/api/courses.js:
```diff
@@ -4,13 +4,93 @@ const router = require('express').Router({
+ * Helper for getting course question data
+ * This filters all present columns in question data returned from the sequelize query to
+ * include only the columns in our list of selectedColumns.
+ * This is to account for the case where a column may be absent or missing from questions.
```
Curious: did you see this case while testing?

jackieo5023 replied:
I didn't, and I guess after thinking about it it's not too important - it would just account for if we took out a column or something (and tests would catch that).  I could take it out if that would make more sense

nwalters512 replied:
Yeah, my hope is that tests would account for something like that (or that we'd catch that in code review). Iterating over every key of potentially tens of thousands of questions for a response seems to be doing a lot of unnecessary work, I think we'd be fine without this!

nwalters512 commented on src/pages/course.js:
```diff
@@ -57,6 +57,14 @@ const Course = props => {
+            <Button
+              color="primary"
+              className="mr-3 mt-3"
+              href={`/api/courses/${props.courseId}/data/questions`}
```
This should use `withBaseUrl` from `util/index.ts` to account for the case where the queue is deployed at a path other than the domain root (like in prod, where it's at `queue.illinois.edu/q/`).

nwalters512 commented on src/api/courses.js:
```diff
@@ -4,13 +4,93 @@ const router = require('express').Router({
+  })
+  csv.unshift(splitHeader.join(','))
+
+  return csv.join('\r\n')
```
Is there a particular reason this is using `\r\n` and not just `\n`? Would be good to have that reasoning documented inline.
